File: isabella.ml/src/canon/confidential_range.ml
(** Confidential range proof helpers
    Derived from the verified Canon/ZK/Confidential_Range theory.

    This module keeps the SDK surface native and deterministic while remaining
    a thin composition layer over Isabelle-exported Canon modules. *)

type scalar_commit_params = Commit_sis.commit_params
type commit_opening = Commit_sis.commit_opening

type range_proof = {
  range_bits : int list list;
  range_comps : int list list;
  range_amount_as : int list list;
  range_amount_zs : int list list;
  range_pair_ass : int list list list;
  range_pair_zss : int list list list;
}

let make_range_proof range_bits range_comps range_amount_as range_amount_zs range_pair_ass range_pair_zss =
  { range_bits; range_comps; range_amount_as; range_amount_zs; range_pair_ass; range_pair_zss }

let one_opening p =
  Commit_sis.make_opening [1] (List.init p.Commit_sis.cp_n2 (fun _ -> 0))

let opening_scale k op =
  Commit_sis.make_opening
    (Listvec.scalar_mult k op.Commit_sis.open_msg)
    (Listvec.scalar_mult k op.Commit_sis.open_rand)

let valid_bit_opening p op =
  Confidential_balance.valid_scalar_commit_params p &&
  Commit_sis.valid_opening p op &&
  Norms.all_bounded op.Commit_sis.open_msg 1

let bit_pair_relation p op_bit op_comp =
  valid_bit_opening p op_bit &&
  valid_bit_opening p op_comp &&
  Confidential_balance.amount_of_opening op_bit +
  Confidential_balance.amount_of_opening op_comp = 1

let rec all_bit_pairs p ops_bits ops_comps =
  match ops_bits, ops_comps with
  | [], [] -> true
  | op_bit :: rest_bits, op_comp :: rest_comps ->
      bit_pair_relation p op_bit op_comp && all_bit_pairs p rest_bits rest_comps
  | _ -> false

let rec weighted_opening p b = function
  | [] -> Commit_sis.make_opening [0] (List.init p.Commit_sis.cp_n2 (fun _ -> 0))
  | op :: ops ->
      let rest = weighted_opening p b ops in
      Commit_sis.make_opening
        (Listvec.vec_add op.Commit_sis.open_msg (Listvec.scalar_mult b rest.Commit_sis.open_msg))
        (Listvec.vec_add op.Commit_sis.open_rand (Listvec.scalar_mult b rest.Commit_sis.open_rand))

let rec weighted_commitment p ck b = function
  | [] -> Confidential_balance.rand_commit p ck (List.init p.Commit_sis.cp_n2 (fun _ -> 0))
  | c :: cs ->
      Zq.vec_mod
        (Listvec.vec_add c (Listvec.scalar_mult b (weighted_commitment p ck b cs)))
        p.Commit_sis.cp_q

let range_amount_opening p op_amount ops_bits =
  let weighted = weighted_opening p 2 ops_bits in
  Commit_sis.make_opening
    (Listvec.vec_sub op_amount.Commit_sis.open_msg weighted.Commit_sis.open_msg)
    (Listvec.vec_sub op_amount.Commit_sis.open_rand weighted.Commit_sis.open_rand)

let range_amount_randomness p op_amount ops_bits =
  (range_amount_opening p op_amount ops_bits).Commit_sis.open_rand

let range_pair_opening p op_bit op_comp =
  let one = one_opening p in
  Commit_sis.make_opening
    (Listvec.vec_sub
       (Listvec.vec_add op_bit.Commit_sis.open_msg op_comp.Commit_sis.open_msg)
       one.Commit_sis.open_msg)
    (Listvec.vec_sub
       (Listvec.vec_add op_bit.Commit_sis.open_rand op_comp.Commit_sis.open_rand)
       one.Commit_sis.open_rand)

let range_pair_randomness p op_bit op_comp =
  (range_pair_opening p op_bit op_comp).Commit_sis.open_rand

let rec range_pair_randomnesses p ops_bits ops_comps =
  match ops_bits, ops_comps with
  | [], [] -> []
  | op_bit :: rest_bits, op_comp :: rest_comps ->
      range_pair_randomness p op_bit op_comp :: range_pair_randomnesses p rest_bits rest_comps
  | _ -> []

let one_commitment p ck =
  Commit_sis.commit ck (one_opening p) p.Commit_sis.cp_q

let range_amount_commitment p ck c_amount c_bits =
  Zq.vec_mod
    (Listvec.vec_sub c_amount (weighted_commitment p ck 2 c_bits))
    p.Commit_sis.cp_q

let range_pair_commitment p ck c_bit c_comp =
  Zq.vec_mod
    (Listvec.vec_sub (Listvec.vec_add c_bit c_comp) (one_commitment p ck))
    p.Commit_sis.cp_q

let rec range_pair_commitments p ck c_bits c_comps =
  match c_bits, c_comps with
  | [], [] -> []
  | c_bit :: rest_bits, c_comp :: rest_comps ->
      range_pair_commitment p ck c_bit c_comp :: range_pair_commitments p ck rest_bits rest_comps
  | _ -> []

let range_relation p ck c_amount op_amount ops_bits ops_comps =
  Confidential_balance.valid_scalar_commit_params p &&
  Confidential_balance.valid_confidential_commit_key p ck &&
  Commit_sis.valid_opening p op_amount &&
  Commit_sis.verify_opening ck op_amount c_amount p.Commit_sis.cp_q &&
  all_bit_pairs p ops_bits ops_comps &&
  Confidential_balance.amount_of_opening op_amount =
    Decomp.recompose 2 (List.map Confidential_balance.amount_of_opening ops_bits)

let rec pow_int base exp =
  if exp <= 0 then 1 else base * pow_int base (exp - 1)

let range_amount_witness_bound p k =
  pow_int 2 k * p.Commit_sis.cp_beta

let range_pair_witness_bound p =
  2 * p.Commit_sis.cp_beta

let valid_range_amount_witness p k r =
  Listvec.valid_vec p.Commit_sis.cp_n2 r &&
  Norms.all_bounded r (range_amount_witness_bound p k)

let valid_range_pair_witness p r =
  Listvec.valid_vec p.Commit_sis.cp_n2 r &&
  Norms.all_bounded r (range_pair_witness_bound p)

let valid_range_mask p gamma y =
  Listvec.valid_vec p.Commit_sis.cp_n2 y &&
  Norms.all_bounded y gamma

let range_amount_response_bound p gamma k challenge =
  gamma + (abs challenge * range_amount_witness_bound p k)

let range_pair_response_bound p gamma challenge =
  gamma + (abs challenge * range_pair_witness_bound p)

let valid_range_challenge = Confidential_balance.valid_balance_challenge

let valid_range_amount_response p gamma k challenge z =
  Listvec.valid_vec p.Commit_sis.cp_n2 z &&
  Norms.all_bounded z (range_amount_response_bound p gamma k challenge)

let valid_range_pair_response p gamma challenge z =
  Listvec.valid_vec p.Commit_sis.cp_n2 z &&
  Norms.all_bounded z (range_pair_response_bound p gamma challenge)

let valid_range_masks p gamma ys =
  List.for_all (valid_range_mask p gamma) ys

let valid_range_pair_responses p gamma challenge zs =
  List.for_all (valid_range_pair_response p gamma challenge) zs

let range_fs_rounds = Confidential_balance.balance_fs_rounds

let range_fs_challenges _p ck c_amount c_bits c_comps a_amounts a_pairss =
  let sum_list = List.fold_left ( + ) 0 in
  let base =
    sum_list (List.concat ck) +
    sum_list c_amount +
    sum_list (List.concat c_bits) +
    sum_list (List.concat c_comps) +
    sum_list (List.concat a_amounts) +
    sum_list (List.concat (List.concat a_pairss))
  in
  Repeated_fs.bool_fs_challenges base

let range_amount_sigma_announcements p ck y_amounts =
  List.map (Confidential_balance.rand_commit p ck) y_amounts

let range_amount_sigma_responses r_amount y_amounts challenges =
  Repeated_fs.sigma_response_rounds Confidential_balance.balance_sigma_respond r_amount y_amounts challenges

let range_sigma_announcements p ck ys =
  List.map (Confidential_balance.rand_commit p ck) ys

let rec range_sigma_responses rs ys challenge =
  match rs, ys with
  | [], [] -> []
  | r :: rest_rs, y :: rest_ys ->
      Confidential_balance.balance_sigma_respond r y challenge ::
      range_sigma_responses rest_rs rest_ys challenge
  | _ -> []

let range_pair_sigma_announcement_rounds p ck y_pairss =
  List.map (range_sigma_announcements p ck) y_pairss

let range_pair_sigma_response_rounds rs y_pairss challenges =
  Repeated_fs.sigma_response_rounds range_sigma_responses rs y_pairss challenges

let canonical_range_challenge _p ck c_amount c_bits c_comps a_amounts a_pairss =
  let sum_list = List.fold_left ( + ) 0 in
  (sum_list (List.concat ck) +
   sum_list c_amount +
   sum_list (List.concat c_bits) +
   sum_list (List.concat c_comps) +
   sum_list (List.concat a_amounts) +
   sum_list (List.concat (List.concat a_pairss))) mod 2

let range_amount_sigma_verify p gamma k ck c a challenge z =
  Confidential_balance.valid_scalar_commit_params p &&
  Confidential_balance.valid_confidential_commit_key p ck &&
  Listvec.valid_vec p.Commit_sis.cp_m a &&
  valid_range_challenge p challenge &&
  valid_range_amount_response p gamma k challenge z &&
  Confidential_balance.rand_commit p ck z =
    Zq.vec_mod (Listvec.vec_add a (Listvec.scalar_mult challenge c)) p.Commit_sis.cp_q

let range_pair_sigma_verify p gamma ck c a challenge z =
  Confidential_balance.valid_scalar_commit_params p &&
  Confidential_balance.valid_confidential_commit_key p ck &&
  Listvec.valid_vec p.Commit_sis.cp_m a &&
  valid_range_challenge p challenge &&
  valid_range_pair_response p gamma challenge z &&
  Confidential_balance.rand_commit p ck z =
    Zq.vec_mod (Listvec.vec_add a (Listvec.scalar_mult challenge c)) p.Commit_sis.cp_q

let rec range_sigma_verify_pairs p gamma ck cs as_ challenge zs =
  match cs, as_, zs with
  | [], [], [] -> true
  | c :: rest_cs, a :: rest_as, z :: rest_zs ->
      range_pair_sigma_verify p gamma ck c a challenge z &&
      range_sigma_verify_pairs p gamma ck rest_cs rest_as challenge rest_zs
  | _ -> false

let range_fs_prove p gamma k ck c_amount op_amount ops_bits ops_comps y_amounts y_pairss =
  let c_bits = List.map (fun op -> Commit_sis.commit ck op p.Commit_sis.cp_q) ops_bits in
  let c_comps = List.map (fun op -> Commit_sis.commit ck op p.Commit_sis.cp_q) ops_comps in
  let r_amount = range_amount_randomness p op_amount ops_bits in
  let r_pairs = range_pair_randomnesses p ops_bits ops_comps in
  let a_amounts = range_amount_sigma_announcements p ck y_amounts in
  let a_pairss = range_pair_sigma_announcement_rounds p ck y_pairss in
  let challenges = range_fs_challenges p ck c_amount c_bits c_comps a_amounts a_pairss in
  let z_amounts = range_amount_sigma_responses r_amount y_amounts challenges in
  let z_pairss = range_pair_sigma_response_rounds r_pairs y_pairss challenges in
  if range_relation p ck c_amount op_amount ops_bits ops_comps &&
     List.length ops_bits = k &&
     List.length ops_comps = k &&
     List.length y_amounts = range_fs_rounds &&
     List.length y_pairss = range_fs_rounds &&
     List.for_all (valid_range_mask p gamma) y_amounts &&
     List.for_all
       (fun ys -> List.length ys = k && valid_range_masks p gamma ys)
       y_pairss &&
     List.for_all2
       (valid_range_amount_response p gamma k)
       challenges
       z_amounts &&
     List.for_all2
       (fun challenge zs ->
         List.length zs = k &&
         valid_range_pair_responses p gamma challenge zs)
       challenges
       z_pairss
  then Some (make_range_proof c_bits c_comps a_amounts z_amounts a_pairss z_pairss)
  else None

let range_fs_verify p gamma k ck c_amount proof =
  let c_bits = proof.range_bits in
  let c_comps = proof.range_comps in
  let a_amounts = proof.range_amount_as in
  let a_pairss = proof.range_pair_ass in
  let z_amounts = proof.range_amount_zs in
  let z_pairss = proof.range_pair_zss in
  let challenges = range_fs_challenges p ck c_amount c_bits c_comps a_amounts a_pairss in
  let c_amount_res = range_amount_commitment p ck c_amount c_bits in
  let c_pair_res = range_pair_commitments p ck c_bits c_comps in
  Listvec.valid_vec p.Commit_sis.cp_m c_amount &&
  List.length c_bits = k &&
  List.length c_comps = k &&
  List.length a_amounts = range_fs_rounds &&
  List.length a_pairss = range_fs_rounds &&
  List.length z_amounts = range_fs_rounds &&
  List.length z_pairss = range_fs_rounds &&
  List.for_all2
    (fun ((a_amount, a_pairs), challenge) (z_amount, z_pairs) ->
      List.length a_pairs = k &&
      List.length z_pairs = k &&
      range_amount_sigma_verify p gamma k ck c_amount_res a_amount challenge z_amount &&
      range_sigma_verify_pairs p gamma ck c_pair_res a_pairs challenge z_pairs)
    (List.combine (List.combine a_amounts a_pairss) challenges)
    (List.combine z_amounts z_pairss)

File: isabella.hs/src/Canon/ZK/Confidential_Range.hs
-- | Canon.ZK.Confidential_Range: native Haskell facade for the
-- confidential-range proof slice exported from Canon.
module Canon.ZK.Confidential_Range
  ( RangeProof(..)
  , makeRangeProof
  , oneOpening
  , openingScale
  , validBitOpening
  , bitPairRelation
  , weightedOpening
  , weightedCommitment
  , rangeAmountOpening
  , rangeAmountRandomness
  , rangePairOpening
  , rangePairRandomness
  , rangePairRandomnesses
  , oneCommitment
  , rangeAmountCommitment
  , rangePairCommitment
  , rangePairCommitments
  , rangeRelation
  , rangeAmountWitnessBound
  , rangePairWitnessBound
  , validRangeAmountWitness
  , validRangePairWitness
  , validRangeMask
  , rangeAmountResponseBound
  , rangePairResponseBound
  , validRangeChallenge
  , validRangeAmountResponse
  , validRangePairResponse
  , rangeFsRounds
  , rangeFsChallenges
  , rangeAmountSigmaAnnouncements
  , rangeAmountSigmaResponses
  , rangePairSigmaAnnouncementRounds
  , rangePairSigmaResponseRounds
  , rangeSigmaAnnouncements
  , rangeSigmaResponses
  , canonicalRangeChallenge
  , rangeAmountSigmaVerify
  , rangePairSigmaVerify
  , rangeSigmaVerifyPairs
  , rangeFsProve
  , rangeFsVerify
  ) where

import qualified Canon.Commit_sis as Commit
import qualified Canon.Confidential_balance as ConfidentialBalance
import qualified Canon.Decomp as Decomp
import qualified Canon.Listvec as Listvec
import qualified Canon.Norms as Norms
import qualified Canon.ZK.Internal.RepeatedFS as RepeatedFS
import qualified Canon.Zq as Zq

data RangeProof = RangeProof
  { range_bits :: [[Int]]
  , range_comps :: [[Int]]
  , range_amount_as :: [[Int]]
  , range_amount_zs :: [[Int]]
  , range_pair_ass :: [[[Int]]]
  , range_pair_zss :: [[[Int]]]
  }
  deriving (Eq, Show)

makeRangeProof :: [[Int]] -> [[Int]] -> [[Int]] -> [[Int]] -> [[[Int]]] -> [[[Int]]] -> RangeProof
makeRangeProof = RangeProof

oneOpening :: Commit.CommitParams -> Commit.CommitOpening
oneOpening p = Commit.makeOpening [1] (replicate (Commit.cp_n2 p) 0)

openingScale :: Int -> Commit.CommitOpening -> Commit.CommitOpening
openingScale k op =
  Commit.makeOpening
    (Listvec.scalar_mult k (Commit.open_msg op))
    (Listvec.scalar_mult k (Commit.open_rand op))

-- Keep the exported relation logic explicit rather than clever.
validBitOpening :: Commit.CommitParams -> Commit.CommitOpening -> Bool
validBitOpening p op =
  ConfidentialBalance.validScalarCommitParams p &&
  Commit.validOpening p op &&
  Norms.all_bounded (Commit.open_msg op) 1

bitPairRelation :: Commit.CommitParams -> Commit.CommitOpening -> Commit.CommitOpening -> Bool
bitPairRelation p opBit opComp =
  validBitOpening p opBit &&
  validBitOpening p opComp &&
  ConfidentialBalance.amountOfOpening opBit + ConfidentialBalance.amountOfOpening opComp == 1

allBitPairs :: Commit.CommitParams -> [Commit.CommitOpening] -> [Commit.CommitOpening] -> Bool
allBitPairs _ [] [] = True
allBitPairs p (opBit : opsBits) (opComp : opsComps) =
  bitPairRelation p opBit opComp && allBitPairs p opsBits opsComps
allBitPairs _ _ _ = False

weightedOpening :: Commit.CommitParams -> Int -> [Commit.CommitOpening] -> Commit.CommitOpening
weightedOpening p _ [] = Commit.makeOpening [0] (replicate (Commit.cp_n2 p) 0)
weightedOpening p b (op : ops) =
  let rest = weightedOpening p b ops
   in Commit.makeOpening
        (Listvec.vec_add (Commit.open_msg op) (Listvec.scalar_mult b (Commit.open_msg rest)))
        (Listvec.vec_add (Commit.open_rand op) (Listvec.scalar_mult b (Commit.open_rand rest)))

weightedCommitment :: Commit.CommitParams -> [[Int]] -> Int -> [[Int]] -> [Int]
weightedCommitment p ck _ [] =
  ConfidentialBalance.randCommit p ck (replicate (Commit.cp_n2 p) 0)
weightedCommitment p ck b (c : cs) =
  Zq.vec_mod
    (Listvec.vec_add c (Listvec.scalar_mult b (weightedCommitment p ck b cs)))
    (Commit.cp_q p)

rangeAmountOpening :: Commit.CommitParams -> Commit.CommitOpening -> [Commit.CommitOpening] -> Commit.CommitOpening
rangeAmountOpening p opAmount opsBits =
  Commit.makeOpening
    (Listvec.vec_sub (Commit.open_msg opAmount) (Commit.open_msg (weightedOpening p 2 opsBits)))
    (Listvec.vec_sub (Commit.open_rand opAmount) (Commit.open_rand (weightedOpening p 2 opsBits)))

rangeAmountRandomness :: Commit.CommitParams -> Commit.CommitOpening -> [Commit.CommitOpening] -> [Int]
rangeAmountRandomness p opAmount opsBits = Commit.open_rand (rangeAmountOpening p opAmount opsBits)

rangePairOpening :: Commit.CommitParams -> Commit.CommitOpening -> Commit.CommitOpening -> Commit.CommitOpening
rangePairOpening p opBit opComp =
  let one = oneOpening p
      sumMsg = Listvec.vec_add (Commit.open_msg opBit) (Commit.open_msg opComp)
      sumRand = Listvec.vec_add (Commit.open_rand opBit) (Commit.open_rand opComp)
   in Commit.makeOpening
        (Listvec.vec_sub sumMsg (Commit.open_msg one))
        (Listvec.vec_sub sumRand (Commit.open_rand one))

rangePairRandomness :: Commit.CommitParams -> Commit.CommitOpening -> Commit.CommitOpening -> [Int]
rangePairRandomness p opBit opComp = Commit.open_rand (rangePairOpening p opBit opComp)

rangePairRandomnesses :: Commit.CommitParams -> [Commit.CommitOpening] -> [Commit.CommitOpening] -> [[Int]]
rangePairRandomnesses _ [] [] = []
rangePairRandomnesses p (opBit : opsBits) (opComp : opsComps) =
  rangePairRandomness p opBit opComp : rangePairRandomnesses p opsBits opsComps
rangePairRandomnesses _ _ _ = []

oneCommitment :: Commit.CommitParams -> [[Int]] -> [Int]
oneCommitment p ck = Commit.commit ck (oneOpening p) (Commit.cp_q p)

rangeAmountCommitment :: Commit.CommitParams -> [[Int]] -> [Int] -> [[Int]] -> [Int]
rangeAmountCommitment p ck cAmount cBits =
  Zq.vec_mod
    (Listvec.vec_sub cAmount (weightedCommitment p ck 2 cBits))
    (Commit.cp_q p)

rangePairCommitment :: Commit.CommitParams -> [[Int]] -> [Int] -> [Int] -> [Int]
rangePairCommitment p ck cBit cComp =
  Zq.vec_mod
    (Listvec.vec_sub (Listvec.vec_add cBit cComp) (oneCommitment p ck))
    (Commit.cp_q p)

rangePairCommitments :: Commit.CommitParams -> [[Int]] -> [[Int]] -> [[Int]] -> [[Int]]
rangePairCommitments _ _ [] [] = []
rangePairCommitments p ck (cBit : cBits) (cComp : cComps) =
  rangePairCommitment p ck cBit cComp : rangePairCommitments p ck cBits cComps
rangePairCommitments _ _ _ _ = []

rangeRelation ::
  Commit.CommitParams ->
  [[Int]] ->
  [Int] ->
  Commit.CommitOpening ->
  [Commit.CommitOpening] ->
  [Commit.CommitOpening] ->
  Bool
rangeRelation p ck cAmount opAmount opsBits opsComps =
  ConfidentialBalance.validScalarCommitParams p &&
  ConfidentialBalance.validConfidentialCommitKey p ck &&
  Commit.validOpening p opAmount &&
  Commit.verify_opening ck opAmount cAmount (Commit.cp_q p) &&
  allBitPairs p opsBits opsComps &&
  ConfidentialBalance.amountOfOpening opAmount ==
    Decomp.recompose 2 (map ConfidentialBalance.amountOfOpening opsBits)

rangeAmountWitnessBound :: Commit.CommitParams -> Int -> Int
rangeAmountWitnessBound p k = (2 ^ k) * Commit.cp_beta p

rangePairWitnessBound :: Commit.CommitParams -> Int
rangePairWitnessBound p = 2 * Commit.cp_beta p

validRangeAmountWitness :: Commit.CommitParams -> Int -> [Int] -> Bool
validRangeAmountWitness p k r =
  Listvec.valid_vec (Commit.cp_n2 p) r &&
  Norms.all_bounded r (rangeAmountWitnessBound p k)

validRangePairWitness :: Commit.CommitParams -> [Int] -> Bool
validRangePairWitness p r =
  Listvec.valid_vec (Commit.cp_n2 p) r &&
  Norms.all_bounded r (rangePairWitnessBound p)

validRangeMask :: Commit.CommitParams -> Int -> [Int] -> Bool
validRangeMask p gamma y =
  Listvec.valid_vec (Commit.cp_n2 p) y &&
  Norms.all_bounded y gamma

rangeAmountResponseBound :: Commit.CommitParams -> Int -> Int -> Int -> Int
rangeAmountResponseBound p gamma k challenge =
  gamma + abs challenge * rangeAmountWitnessBound p k

rangePairResponseBound :: Commit.CommitParams -> Int -> Int -> Int
rangePairResponseBound p gamma challenge =
  gamma + abs challenge * rangePairWitnessBound p

validRangeChallenge :: Commit.CommitParams -> Int -> Bool
validRangeChallenge = ConfidentialBalance.validBalanceChallenge

validRangeAmountResponse :: Commit.CommitParams -> Int -> Int -> Int -> [Int] -> Bool
validRangeAmountResponse p gamma k challenge z =
  Listvec.valid_vec (Commit.cp_n2 p) z &&
  Norms.all_bounded z (rangeAmountResponseBound p gamma k challenge)

validRangePairResponse :: Commit.CommitParams -> Int -> Int -> [Int] -> Bool
validRangePairResponse p gamma challenge z =
  Listvec.valid_vec (Commit.cp_n2 p) z &&
  Norms.all_bounded z (rangePairResponseBound p gamma challenge)

validRangeMasks :: Commit.CommitParams -> Int -> [[Int]] -> Bool
validRangeMasks p gamma = all (validRangeMask p gamma)

validRangePairResponses :: Commit.CommitParams -> Int -> Int -> [[Int]] -> Bool
validRangePairResponses p gamma challenge = all (validRangePairResponse p gamma challenge)

rangeFsRounds :: Int
rangeFsRounds = ConfidentialBalance.balanceFsRounds

rangeFsChallenges ::
  Commit.CommitParams ->
  [[Int]] ->
  [Int] ->
  [[Int]] ->
  [[Int]] ->
  [[Int]] ->
  [[[Int]]] ->
  [Int]
rangeFsChallenges _p ck cAmount cBits cComps aAmounts aPairss =
  let challengeBase =
        sum (concat ck) +
        sum cAmount +
        sum (concat cBits) +
        sum (concat cComps) +
        sum (concat aAmounts) +
        sum (concat (concat aPairss))
   in RepeatedFS.boolFsChallenges challengeBase

rangeAmountSigmaAnnouncements :: Commit.CommitParams -> [[Int]] -> [[Int]] -> [[Int]]
rangeAmountSigmaAnnouncements p ck = map (ConfidentialBalance.randCommit p ck)

rangeAmountSigmaResponses :: [Int] -> [[Int]] -> [Int] -> [[Int]]
rangeAmountSigmaResponses rAmount yAmounts challenges =
  zipWith (ConfidentialBalance.balanceSigmaRespond rAmount) yAmounts challenges

rangePairSigmaAnnouncementRounds ::
  Commit.CommitParams ->
  [[Int]] ->
  [[[Int]]] ->
  [[[Int]]]
rangePairSigmaAnnouncementRounds p ck = map (rangeSigmaAnnouncements p ck)

rangePairSigmaResponseRounds ::
  [[Int]] ->
  [[[Int]]] ->
  [Int] ->
  [[[Int]]]
rangePairSigmaResponseRounds rs yPairss challenges =
  RepeatedFS.sigmaResponseRounds rangeSigmaResponses rs yPairss challenges

rangeSigmaAnnouncements :: Commit.CommitParams -> [[Int]] -> [[Int]] -> [[Int]]
rangeSigmaAnnouncements p ck = map (ConfidentialBalance.randCommit p ck)

rangeSigmaResponses :: [[Int]] -> [[Int]] -> Int -> [[Int]]
rangeSigmaResponses [] [] _ = []
rangeSigmaResponses (r : rs) (y : ys) challenge =
  ConfidentialBalance.balanceSigmaRespond r y challenge : rangeSigmaResponses rs ys challenge
rangeSigmaResponses _ _ _ = []

canonicalRangeChallenge ::
  Commit.CommitParams ->
  [[Int]] ->
  [Int] ->
  [[Int]] ->
  [[Int]] ->
  [[Int]] ->
  [[[Int]]] ->
  Int
canonicalRangeChallenge p ck cAmount cBits cComps aAmounts aPairss =
  mod
    ( sum (concat ck)
    + sum cAmount
    + sum (concat cBits)
    + sum (concat cComps)
    + sum (concat aAmounts)
    + sum (concat (concat aPairss))
    )
    2

rangeAmountSigmaVerify ::
  Commit.CommitParams ->
  Int ->
  Int ->
  [[Int]] ->
  [Int] ->
  [Int] ->
  Int ->
  [Int] ->
  Bool
rangeAmountSigmaVerify p gamma k ck c a challenge z =
  ConfidentialBalance.validScalarCommitParams p &&
  ConfidentialBalance.validConfidentialCommitKey p ck &&
  Listvec.valid_vec (Commit.cp_m p) a &&
  validRangeChallenge p challenge &&
  validRangeAmountResponse p gamma k challenge z &&
  ConfidentialBalance.randCommit p ck z ==
    Zq.vec_mod (Listvec.vec_add a (Listvec.scalar_mult challenge c)) (Commit.cp_q p)

rangePairSigmaVerify ::
  Commit.CommitParams ->
  Int ->
  [[Int]] ->
  [Int] ->
  [Int] ->
  Int ->
  [Int] ->
  Bool
rangePairSigmaVerify p gamma ck c a challenge z =
  ConfidentialBalance.validScalarCommitParams p &&
  ConfidentialBalance.validConfidentialCommitKey p ck &&
  Listvec.valid_vec (Commit.cp_m p) a &&
  validRangeChallenge p challenge &&
  validRangePairResponse p gamma challenge z &&
  ConfidentialBalance.randCommit p ck z ==
    Zq.vec_mod (Listvec.vec_add a (Listvec.scalar_mult challenge c)) (Commit.cp_q p)

rangeSigmaVerifyPairs ::
  Commit.CommitParams ->
  Int ->
  [[Int]] ->
  [[Int]] ->
  [[Int]] ->
  Int ->
  [[Int]] ->
  Bool
rangeSigmaVerifyPairs _ _ _ [] [] _ [] = True
rangeSigmaVerifyPairs p gamma ck (c : cs) (a : as) challenge (z : zs) =
  rangePairSigmaVerify p gamma ck c a challenge z &&
  rangeSigmaVerifyPairs p gamma ck cs as challenge zs
rangeSigmaVerifyPairs _ _ _ _ _ _ _ = False

rangeFsProve ::
  Commit.CommitParams ->
  Int ->
  Int ->
  [[Int]] ->
  [Int] ->
  Commit.CommitOpening ->
  [Commit.CommitOpening] ->
  [Commit.CommitOpening] ->
  [[Int]] ->
  [[[Int]]] ->
  Maybe RangeProof
rangeFsProve p gamma k ck cAmount opAmount opsBits opsComps yAmounts yPairss =
  let cBits = map (\op -> Commit.commit ck op (Commit.cp_q p)) opsBits
      cComps = map (\op -> Commit.commit ck op (Commit.cp_q p)) opsComps
      rAmount = rangeAmountRandomness p opAmount opsBits
      rPairs = rangePairRandomnesses p opsBits opsComps
      aAmounts = rangeAmountSigmaAnnouncements p ck yAmounts
      aPairss = rangePairSigmaAnnouncementRounds p ck yPairss
      challenges = rangeFsChallenges p ck cAmount cBits cComps aAmounts aPairss
      zAmounts = rangeAmountSigmaResponses rAmount yAmounts challenges
      zPairss = rangePairSigmaResponseRounds rPairs yPairss challenges
   in if rangeRelation p ck cAmount opAmount opsBits opsComps &&
         length opsBits == k &&
         length opsComps == k &&
         length yAmounts == rangeFsRounds &&
         length yPairss == rangeFsRounds &&
         and [validRangeMask p gamma (yAmounts !! i) | i <- [0 .. rangeFsRounds - 1]] &&
         and
           [ length (yPairss !! i) == k &&
             validRangeMasks p gamma (yPairss !! i)
           | i <- [0 .. rangeFsRounds - 1]
           ] &&
         and
           [ validRangeAmountResponse p gamma k (challenges !! i) (zAmounts !! i)
           | i <- [0 .. rangeFsRounds - 1]
           ] &&
         and
           [ length (zPairss !! i) == k &&
             validRangePairResponses p gamma (challenges !! i) (zPairss !! i)
           | i <- [0 .. rangeFsRounds - 1]
           ]
        then Just (RangeProof cBits cComps aAmounts zAmounts aPairss zPairss)
        else Nothing

rangeFsVerify :: Commit.CommitParams -> Int -> Int -> [[Int]] -> [Int] -> RangeProof -> Bool
rangeFsVerify p gamma k ck cAmount proof =
  let cBits = range_bits proof
      cComps = range_comps proof
      aAmounts = range_amount_as proof
      aPairss = range_pair_ass proof
      zAmounts = range_amount_zs proof
      zPairss = range_pair_zss proof
      challenges = rangeFsChallenges p ck cAmount cBits cComps aAmounts aPairss
      cAmountRes = rangeAmountCommitment p ck cAmount cBits
      cPairRes = rangePairCommitments p ck cBits cComps
   in Listvec.valid_vec (Commit.cp_m p) cAmount &&
      length cBits == k &&
      length cComps == k &&
      length aAmounts == rangeFsRounds &&
      length aPairss == rangeFsRounds &&
      length zAmounts == rangeFsRounds &&
      length zPairss == rangeFsRounds &&
      and
        [ length (aPairss !! i) == k &&
          length (zPairss !! i) == k &&
          rangeAmountSigmaVerify p gamma k ck cAmountRes (aAmounts !! i) (challenges !! i) (zAmounts !! i) &&
          rangeSigmaVerifyPairs p gamma ck cPairRes (aPairss !! i) (challenges !! i) (zPairss !! i)
        | i <- [0 .. rangeFsRounds - 1]
        ]
